[api] Tags enhancements

imtayadeway opened the pull request:
Resolves https://trello.com/c/xsN82YoU

@abellotti please review

imtayadeway commented:
@miq-bot add-label api, core, enhancement

imtayadeway commented:
Hmm, not sure why the build failed. The failing test passes for me locally

abellotti commented on app/controllers/api_controller/tags.rb:
```diff
@@ -70,12 +70,12 @@ def edit_resource_tags(type, id, data)
         _, category_id = parse_href(data["href"])
       else
-        raise BadRequestError, "Category id or href needs to be specified for creating a new tag resource"
+        raise BadRequestError, "Category id, href or name needs to be specified for creating a new tag resource"
```
could also be written as:
```
category_id = parse_id(data, :categories) || parse_by_attr(data, :categories, %w(name))
raise BadRequestError, "Category id, href or name needs to be specified for creating a new tag resource" unless category_id
```

abellotti commented:
Are you solving the adding/removing tags by subcollection mention in the trello in this PR ?
i.e. POST /api/categories/:id/tags    actions "create" and "delete" 

Thanks.

imtayadeway commented:
@abellotti creating tags by subcollection was added in the original PR at the last minute. Didn't realize we'd want to delete too - I'll add that in this PR

abellotti commented:
Ok, just tried with subcollection and it worked great. Thanks !!

The one thing I noticed, let's say I add 2 tags via subcollection:

POST /api/categories/:id/tags

```{
  "action" : "create",
  "resources" : [ 
    { "name" : "sub_tag_n11", "description" : "Sub Tag n11" },
    { "name" : "sub_tag_n12", "description" : "Sub Tag n12" }
  ]
}
```

Then I delete those 2 new tags via api delete tags/:id  (2 new ids returned).

I cannot recreate the tags with above request. I get the following:

```
{
  "error": {
    "kind": "bad_request",
    "message": "Could not create a new tag - Description has already been taken",
    "klass": "ApiController::BadRequestError"
  }
}
```

Maybe the delete needs to delete the Classification entry for the tag too ?  not sure.

Thanks,

imtayadeway commented:
@abellotti hm, I'll investigate. I wonder if we should delete the tag's classification data at the model level, since this is essentially metadata on the tag. @gtanzillo unless I'm mistaken?

What's the expected usecase for doing the "delete" action? To delete all of a category's tags? Or to provide the tag name, id, etc... in the POST body?

abellotti commented:
The deletes on subcollection would allow a user to delete one or more tags of that category, by name, href or id.

i.e.
POST /api/categories/:id/tags

```
{
  "action" : "delete",
  "resources" : [ 
    { "name" : "/managed/some_path/sub_tag_n11" },
    { "id" : 1492 },
    ...
  ]
}
```

Thanks

imtayadeway commented:
@abellotti thanks! Just made some updates which I think take care of all concerns.
@gtanzillo can you review?

gtanzillo commented on app/controllers/api_controller/tags.rb:
```diff
@@ -34,6 +34,14 @@ def tags_create_resource(parent, _type, _id, data)
+    def tags_delete_resource(_parent, _type, _id, data)
+      tag_id = parse_id(data, :tags) || parse_by_attr(data, :tags, %w(name))
+      raise BadRequestError, "Tag id, href or name needs to be specified for deleting a tag resource" unless tag_id
+      tag = Tag.find_by_id(tag_id)
```
@imtayadeway You'll need to delete the Classification entries here instead of just the tag. There is a before_destroy that calls delete_tags_and_entries. That method deletes the corresponding tags and removes the taggings on any objects that were tagged with the entry.

imtayadeway commented:
@gtanzillo thanks! I implemented a custom `#delete_resource_tags` method which sends the destroy to the tag's classification and reverted the `Tag` change.

abellotti commented on app/controllers/api_controller/tags.rb:
```diff
@@ -67,15 +75,18 @@ def edit_resource_tags(type, id, data)
 
+    def delete_resource_tags(_type, id, *)
+      entry = Classification.find_by_tag_id(id)
+      raise BadREquestError, "Failed to find tag resource" unless entry
```
BadREquestError => BadRequestError.

abellotti commented on app/controllers/api_controller/tags.rb:
```diff
@@ -67,15 +75,18 @@ def edit_resource_tags(type, id, data)
+      entry = Classification.find_by_tag_id(id)
+      raise BadREquestError, "Failed to find tag resource" unless entry
+      entry.destroy
+    end
```
we'll need to return action signature, similar to what is done in delete_resource_action(klass, type, id).

imtayadeway commented:
@abellotti just applied your recommendations - let me know if there's anything else here and I'll make sure to squash before merge. Thanks!

imtayadeway commented:
@gtanzillo I decided to squash down into 7 commits, as I felt each one was addressing a separate concern. Should be GTG when (/if) it turns green

abellotti commented:
LGTM !!   :cake:
